Concepts/ClusterAdministration: Expand Node Autoscaling documentation

towca opened the pull request:
Change Cluster Autoscaling concept page to Node Autoscaling, and expand it in accordance to the Cluster Autoscaler/Karpenter Alignment AEP: https://docs.google.com/document/d/1rHhltfLV5V1kcnKr_mKRKDC4ZFPYGP4Tde2Zy-LE72w/edit#heading=h.iof64m6gewln.

The documentation tentatively mentions one of the workload-level APIs we're supposed to align on (see https://github.com/kubernetes/autoscaler/issues/6648). We can comment that section out until we align and implement the new version in both CA and Karpenter.

Tracking issue: https://github.com/kubernetes/autoscaler/issues/6646

Staged preview: https://deploy-preview-45802--kubernetes-io-main-staging.netlify.app/docs/concepts/cluster-administration/node-autoscaling/

cc @jonathan-innis @MaciekPytel @gjtempleton WDYT?

towca commented:
/hold
/assign @jonathan-innis 
/assign @gjtempleton 
/assign @MaciekPytel

jackfrancis commented on content/en/docs/concepts/cluster-administration/node-autoscaling.md:
```diff
@@ -0,0 +1,264 @@
+with a particular set of pods, lack of cloud provider capacity).
+
+{{< note >}}
+Only the scheduling constraints of pods (for example resource requests, node selectors) are taken into account when determining Nodes to
```
I think this statement "Only the scheduling constraints of pods... are taken into account when determining Nodes to provision." is not entirely correct for Karpenter. Maybe a better way to put it would be:

"The fundamental input for determining node provisioning is pod scheduling constraints."

This would leave open other node provisioning inputs (such as vm cost). I don't think we have to state this, but I think saying _Only_ the scheduling constraints of pods... is too strong and excludes other node autoprovisioning considerations.

cc @jonathan-innis

MaciekPytel replied:
I think the intent was to emphasize that real resource usage is not really considered and neither CA nor Karpenter will do anything if your pod is OOMing. I think that is worth emphasizing.
But I agree that the wording is misleading - both Karpenter and CA look at VM price (in case of CA only if you enable price expander). And I could imagine a lot of other signals - e.g. spot availability. Let's rephrase this.

jackfrancis commented on content/en/docs/concepts/cluster-administration/node-autoscaling.md:
```diff
@@ -0,0 +1,264 @@
+to autoscale Nodes based on actual resource usage, you can [combine workload and Node autoscaling](#multi-scaling).
+{{< /note >}}
+
+Depending on the choice of autoscaler, provisioning can also be triggered by
```
I think minimum (or maximum) node thresholds are more of an autoscaler configuration rather than a required behavior of a particular autoscaler. So I'd probably replace "Depending on the choice of autoscaler..." with

"Autoscaler configuration may also include other node provisioning triggers (for example the number of Nodes falling below a configured minimum limit)."

jackfrancis commented on content/en/docs/concepts/cluster-administration/node-autoscaling.md:
```diff
@@ -0,0 +1,264 @@
+these Pods pending. This scenario is usually automatically recovered by provisioning.
+{{< /note >}}
+
+Depending on the choice of autoscaler, consolidation can also be triggered by
```
Again I would probably change this to "Autoscaler configuration may also enable triggering consolidation by other conditions..."

jackfrancis commented on content/en/docs/concepts/cluster-administration/node-autoscaling.md:
```diff
@@ -0,0 +1,264 @@
+
+### Autoscaler implementations
+
+Cluster autoscaler and Karpenter are the two Node autoscalers currently supported by
```
is "sponsored by SIG Autoscaling" better than "supported by SIG Autoscaling"?

jackfrancis commented on content/en/docs/concepts/cluster-administration/node-autoscaling.md:
```diff
@@ -0,0 +1,264 @@
+experience. Both will provision new Nodes for unschedulable Pods, and both will
+consolidate the Nodes that are no longer optimally utilized.
+
+The most of the difference between the autoscalers is in how they're set up and
```
I think we've already stated the first part of this sentence already in the doc. We could add a simple statement here "Different autoscalers may provide features outsice the Node autoscaling scope described on this page, and those additional features may differ between them."

jackfrancis commented:
@Bryce-Soghigian @tallaxes because AKS supports both autoscalers it would be good to have your perspective on how well this document reflects those differences. Essentially, do the customer-facing docs that help AKS customers choose one or the other agree with the concepts in this document? (they should!)

Bryce-Soghigian commented on content/en/docs/concepts/cluster-administration/node-autoscaling.md:
```diff
@@ -0,0 +1,264 @@
+
+Additional context:
+* Documentation: https://karpenter.sh/
+* Supported cloud providers: TBD
```
```suggestion
* Supported cloud providers: https://github.com/kubernetes-sigs/karpenter?tab=readme-ov-file#supported-cloudproviders
```

sftim commented on content/en/docs/concepts/cluster-administration/cluster-autoscaling.md:
```diff
@@ -1,117 +0,0 @@
-consolidate Pods onto a smaller number of nodes, to help with automatic scale down
-when the cluster has space capacity.
-
-### Sizing a workload based on cluster size
-
-#### Cluster proportional autoscaler
-
-For workloads that need to be scaled based on the size of the cluster (for example
-`cluster-dns` or other system components), you can use the
-[_Cluster Proportional Autoscaler_](https://github.com/kubernetes-sigs/cluster-proportional-autoscaler).<br />
-
-The Cluster Proportional Autoscaler watches the number of schedulable nodes
-and cores, and scales the number of replicas of the target workload accordingly.
-
-#### Cluster proportional vertical autoscaler
-
-If the number of replicas should stay the same, you can scale your workloads vertically according to the cluster size using
-the [_Cluster Proportional Vertical Autoscaler_](https://github.com/kubernetes-sigs/cluster-proportional-vertical-autoscaler).
-This project is in **beta** and can be found on GitHub.
-
-While the Cluster Proportional Autoscaler scales the number of replicas of a workload, the Cluster Proportional Vertical Autoscaler
-adjusts the resource requests for a workload (for example a Deployment or DaemonSet) based on the number of nodes and/or cores
-in the cluster.
```
If removing this content from the page about node autoscaling, please move it somewhere else (don't drop it).

Bryce-Soghigian replied:
Good catch!

towca replied:
Both of the components mentioned in this section autoscale workloads, not Nodes. Not sure why they were mentioned here in the first place, they're already documented on the workload autoscaling page: https://kubernetes.io/docs/concepts/workloads/autoscaling/#autoscaling-based-on-cluster-size.

sftim replied:
They're mentioned because a reader who is setting up autoscaling might be interested to know that there is code to size workloads based on cluster size.

(in the scenario where we don't mention them, either readers don't learn that those components exist, or they try a search engine and get lucky)

towca replied:
I see your point, but the content on both pages was pretty similar, and seemed a bit redundant. Added a section for node-count-based workload autoscalers in the "Related components" section, linking to the workload autoscaling page to reduce the redundancy. WDYT?

sftim replied:
Yes, this works.

Bryce-Soghigian commented on content/en/docs/concepts/cluster-administration/node-autoscaling.md:
```diff
@@ -0,0 +1,264 @@
+Main differences vs Karpenter:
+* Cluster Autoscaler (CA) provides features related to just Node autoscaling. Node lifecycle
+  beyond that is not in its scope.
+* CA doesn't support auto-provisioning, the Node groups it can provision from have to be
```
Rather than "auto-provisioning" maybe mixed sku autoscaling makes more sense.  

We could also do a small writeup of the advantages of mixed sku autoscaling. Here are some examples that came to my mind when trying to write up some advantages to having access to all skus. Someone can perhaps word it better than me.
----------

1. "75% of scale up failures for Azure's Cluster Autoscaler come from lack of configured quota for the 4 or so sizes customers configure, with less than 1% of customers configuring CAS with more than 4 vm sizes on aks." With mixed sku autoscaling, you aren't limited to those 4 vm sizes. 

2. The per nodeclaim model vs the per nodegroup model allows for different flexibility in the retry model. 

2a. Karpenter puts the responsibility on the cloud provider to select the SKU. Meaning it can choose to use factors like zonal availability, quota, price, and allocation errors in its VM SKU selection decision. 

2b. CAS can't self mitigate the quota errors seen in scale up, but karpenter can retry with a different SKU that has quota or a different zone if there are zonal allocation failures. So retrying continuously in karpenter is less likely to fail.  

These are some of the benefits of having a mixed SKU model, and a per vm retry model. We can retry more intelligently caching skus as unavailable due to zonal, or quota related reasons. 

Another large factor to consider is the pricing selection, being given all SKUs on a given cloudprovider, allows for much better cost savings. CAS has the pricing expander, but the real weakness is that it doesn't have 1000 sizes to pick from. It just can pick based on the current VM Sizes configured by the customer. (As stated earlier, is often no more than 4)

elmiko replied:
great details @Bryce-Soghigian , i'm not so comfortable about using "SKU" so much, as many people are using on-prem and bare metal deployments which look different from public cloud offerings. that said, i totally get what you are talking about.

@jackfrancis this is a tough one for me, because at the core, CAS needs to be informed about the pools it can scale, whereas karpenter can be given some recipes on how to find instances to create pools. scale from zero might appear like autoprovisioning, but the big difference for me is that you still need to configure the node group and link it to an instance type. in cluster-api the user would need to create all the various infrastructure machine templates to be able to have those instance types represented. but on karpenter that expression would be much more compact.

for me, the part to clarify is that CAS will only be able to scale groups that it knows about but karpenter can discover new instances by the constraints it is given. assuming we want to clarify this more.

perhaps we could use a section that describes the difference between "scaling" as opposed to "provisioning" ?

njtran replied:
IIUC, it seems the core difference between the two instance-shape calculations is that Karpenter has more attribute based selection features, and CAS does more instance type matching. Karpenter's NodePools are defaulted to fully inclusive (default = *) where IIUC, CAS is defaulted to fully exclusive, where no instance types in a node group will not be compatible with any pod.

MaciekPytel replied:
> CAS has the pricing expander, but the real weakness is that it doesn't have 1000 sizes to pick from.

In GKE (which is where I have the data on) we see many clusters with >100 nodepools (GKE nodepools; they can be thought of as equivalent of CAS nodegroups). That's enough to effectively give you full spectrum of shapes for optimization (GCE has less instance types than AWS; 100-150 is enough to cover full spectrum of general purpose instances). So even with CAS you can have the SKU selection and zonal handling. 

A major difference is that you need to go through the process of setting all those nodegroups and this needs to be done outside of CAS. With auto-provisioning you no longer have to do any external setup and you get all the different shapes with just a simple configuration of autoscaler.

Bryce-Soghigian replied:
Yep that makes a lot of sense. My point was more from the azure perspective we do not see many clusters creating that many nodepools/nodegroups. GKE has a wonderful CAS integration the rest of us need to strive towards. 

I would also argue at a certain point having 1000 vms that all do the same thing isn't a strength and that 150 or so sizes is more than enough for all intents and purposes.

towca replied:
Judging by this discussion, I think we should probably expand the "Node auto-provisioning" section under "Provisioning", and then this part will be clear?

@njtran's analogy is interesting, but I'm not sure if it maps onto CAS. I see the 2 modes like this, but perhaps I'm missing some Karpenter context:

* Regular provisioning: The user defines a set of homogenous cloud provider VM groups. The groups allow for adding and removing VMs. Based on the pending pods, autoscaler chooses two parameters for provisioning: which group, how many VMs to add. For the autoscaler, this is a well-scoped problem: among the VM options that the user pre-configured, figure out how many new VMs would be needed, then choose the "best" <group, node_count> option, and provision that many Nodes from that group.
* Auto-provisioning: The user doesn't have to define an explicit cloud provider VM group for every possible configuration that could be needed for their workloads. The user only defines a set of constraints on the VMs that can be provisioned. Based on the pending pods, and on the set of constraints, the autoscaler chooses two parameters: the VM config, how many VMs to add. Choosing a VM config from the vast space of possible cloud provider configs is IMO a fundamentally different problem than just selecting between pre-configured configs.

> One could argue that "scale from zero" is functionally equivalent to "auto-provisioning".

@jackfrancis I would disagree here, there's no reason why the regular provisioning model I described above shouldn't work with scale-from-zero. Even while scaling from >0, the autoscaler has to predict how a new Node would look like if a new VM was added to a given group. This prediction is usually straightforward if there are existing Nodes in the group - just take the existing Node and sanitize it. If there are no Nodes in a given group, the prediction has to come from the cloud provider. So some cloud provider integrations will support it while some others won't, but IMO this is more of an implementation detail of a given cloud provider integration for regular provisioning.

From another perspective, the only thing that the autoscaler has to solve on top of scale-from-not-zero to have scale-from-zero is passing "VM template" information from the cloud provider VM group. This information should be available in the group in order to create new VMs, so this should only be the matter of some plumbing and then it's the same as from-not-zero. Contrast this with auto-provisioning where there's no pre-configured template that can be leveraged.

Bryce-Soghigian commented on content/en/docs/concepts/cluster-administration/node-autoscaling.md:
```diff
@@ -0,0 +1,264 @@
+  constraints for the provisioned Nodes instead of pre-configuring Node groups explicitly.
+
+Additional context:
+* Documentation: https://karpenter.sh/
```
This documentation is still not very cloud neutral. Its the best we have now but would be good to revisit this later 

cc: @jonathan-innis, @njtran, @dtzar

njtran replied:
I think @chrisnegus and @dtzar are tackling this, we should definitely update this link with however this effort lands

njtran replied:
This is already being worked on. Personally, I don't think we need to have a TODO here, unless we think that it'll actually help make sure this is updated.

jonathan-innis replied:
Also, I'd imagine that we may end up using the same domain down the line anyways since I don't imagine that we want to shift all of the current users away from the current domain, just extend it

chrisnegus replied:
We are working on making most of the Karpenter.sh site generic, with separate, well-defined sections for supported cloud providers.

Bryce-Soghigian commented on content/en/docs/concepts/cluster-administration/node-autoscaling.md:
```diff
@@ -0,0 +1,264 @@
+always selects specific Nodes to remove as opposed to just resizing the underlying VM group.
+
+Main differences vs Karpenter:
+* Cluster Autoscaler (CA) provides features related to just Node autoscaling. Node lifecycle
```
Might be worth calling out the massive cloud provider support CAS has! Karpenter has a lot of work to do to catch up in that regard!

Bryce-Soghigian commented on content/en/docs/concepts/cluster-administration/node-autoscaling.md:
```diff
@@ -0,0 +1,264 @@
+that best fits the requests of pending Pods. When consolidating, Cluster Autoscaler
+always selects specific Nodes to remove as opposed to just resizing the underlying VM group.
+
+Main differences vs Karpenter:
```
Maybe this belongs in its own section, "CAS VS Karpenter" after the karpenter section?

elmiko commented on content/en/docs/concepts/cluster-administration/node-autoscaling.md:
```diff
@@ -0,0 +1,264 @@
+It works directly with VMs and doesn't rely on provider-side VM groups.
+
+Main differences vs Cluster Autoscaler (CA):
+* In addition to Node autoscaling, Karpenter provides features intended for managing
```
does it make sense to use some of the karpenter terminology here like "disruption"?

elmiko reviewed (commented):
i think the content here looks nice, thanks for proposing it @towca . i have a few questions and comments.

njtran commented on content/en/docs/concepts/cluster-administration/node-autoscaling.md:
```diff
@@ -0,0 +1,264 @@
+### Configure a workload not to be disrupted by consolidation
+{{< feature-state state="stable" for_k8s_version="v1.31" >}}
+
+You can use annotation `autoscaling.kubernetes.io/do-not-disrupt: true` on a Pod to
+prevent it being disrupted by consolidation. Consolidation should never remove a Node
+where a Pod annotated with `autoscaling.kubernetes.io/do-not-disrupt: true` is running.
```
Is this aspirational? Or is this the current CAS annotation that's respected? This is definitely not the annotation that we use in Karpenter today. Maybe we should include the `karpenter.sh/do-not-disrupt` annotation in here until we align on those if CAS has already made the migration?

sftim commented on content/en/docs/concepts/cluster-administration/node-autoscaling.md:
```diff
@@ -0,0 +1,264 @@
+The main component of the cost of running a cluster is usually the cost of the
+VMs backing its Nodes. In order to minimize this cost and still be able to
+run all Pods in the cluster, the Pods should utilize as much of the Nodes'
+resources as possible. From this perspective, the overall Node utilization in
```
You can use normal text, and you can use `---` (on a line by itself) within one note as a divider.

towca commented:
Thanks for all the comments @njtran @elmiko @Bryce-Soghigian @jackfrancis @sftim! I think I've either addressed or responded to everything, please let me know if I missed anything.

sftim commented on content/en/docs/concepts/cluster-administration/node-autoscaling.md:
```diff
@@ -0,0 +1,276 @@
+
+### Balancer
+
+The [balancer](https://github.com/kubernetes/autoscaler/blob/master/balancer/proposals/balancer.md)
+is a component providing horizontal workload autoscaling, with the goal of balancing the number of
+workload replicas between topologies. It can be [combined with Node autoscaling](#multi-scaling) to
+achieve highly-available, cost-efficient workloads.
```
We avoid linking to KEPs in lieu of documentation and especially I wouldn't link to https://github.com/kubernetes/autoscaler/blob/master/balancer/proposals/balancer.md which says it's a KEP but doesn't look like what most contributors would call a KEP.

The best option is to add a new page about the topology balancer into the docs, and then link there.

sftim replied:
I'd remove it TBH.

towca replied:
Removed.

sftim commented on content/en/docs/concepts/cluster-administration/node-autoscaling.md:
```diff
@@ -0,0 +1,276 @@
+and [_Cluster Proportional Vertical
+Autoscaler_](https://github.com/kubernetes-sigs/cluster-proportional-vertical-autoscaler) provide
+horizontal, and vertical workload autoscaling based on the number of Nodes in the cluster. You can
+read more about them [here](/docs/concepts/workloads/autoscaling#autoscaling-based-on-cluster-size).
```
_avoid “here” in hyperlinks_
```suggestion
read more in
[autoscaling based on cluster size](/docs/concepts/workloads/autoscaling#autoscaling-based-on-cluster-size).
```

sftim commented on content/en/docs/concepts/cluster-administration/node-autoscaling.md:
```diff
@@ -0,0 +1,276 @@
+
+### Descheduler
+
+The [descheduler](https://github.com/kubernetes-sigs/descheduler) is a component providing just Node
+consolidation functionality based on custom policies.
```
The descheduler supports more than just node consolidation. For example, it will delete Pods that are restarting too many times. The relevance to node autoscaling is usually to support consolidation; `RemovePodsViolatingTopologySpreadConstraint` is also useful because it might allow recovery from a zone failure, by triggering Pod replacement and indirectly node provisioning once nodes can run in that zone.

sftim commented on content/en/docs/concepts/cluster-administration/node-autoscaling.md:
```diff
@@ -0,0 +1,276 @@
+  to be pre-configured. Karpenter supports auto-provisioning, so the user only has to configure a
+  set of constraints for the provisioned Nodes instead of fully configuring homogenous groups.
+* Cluster Autoscaler provides integrations with numerous cloud providers, including smaller and less
+  popular providers. Karpenter currently only provides integrations with AWS and Azure.
```
Maybe:
```suggestion
  popular providers. The Kubernetes project does not directly publish any autoscaler solution based
  on Karpenter, but third-party releases include integrations with AWS and Azure.
```

sftim replied:
Karpenter, within Kubernetes, is a library you can use to build a node autoscaler. Have I got that right?

njtran replied:
I can see what you're saying @sftim, but I think it's probably fair to just talk about Karpenter as it's been commonly implemented in cloud providers. Specifically in response to Tim's suggestion, here's what I'd propose:

```
* For Cluster Autoscaler, the Kubernetes project directly provides standalone Cloud Provider solutions. For Karpenter, the Kubernetes project publishes Karpenter as a core library of controllers with a common interface which is integrated and implemented by Cloud Providers to build a node autoscaler. Well known implementations include AWS and Azure.
```

sftim replied:
We need to make sure that the changes we merge don't conflict with requirements from our [content guide](https://kubernetes.io/docs/contribute/style/content-guide/).

**Also**, it's good to have content that is timeless. “Karpenter currently only provides integrations with AWS and Azure.” will go stale as soon as someone publishes a third integration.

---

I tweaked the text from https://github.com/kubernetes/website/pull/45802#discussion_r1588403053; @njtran WDYT?

```markdown
* For Cluster Autoscaler, the Kubernetes project directly provides standalone cloud provider
  integrations.
  For Karpenter, the Kubernetes project publishes Karpenter as a core library of controllers with
  a common interface that cloud providers can integrate and extend to build a node autoscaler.
  The implementations of Karpenter node autoscaling include AWS and Azure.
```

We don't limit ourselves to mentioning well known implementations; either we mention no third parties, or even a tiny cloud provider with four customers is welcome to mention their tool.

---

As a personal opinion, it'd be nice to see the CNCF landscape gain the ability to Kubernetes addons (perhaps after running a certification check to make sure they are compatible with stable Kubernetes?)

jonathan-innis replied:
Perhaps to add to this suggestion, we link out to the AWS and Azure implementations that we mention here?

towca replied:
Simplified the suggestions from @sftim and @njtran, added links as @jonathan-innis suggested. WDYT about the current version?

sftim replied:
There's still a “currently”. Can we make this timeless and also avoid the risk that this goes stale due to an external change, such as someone using Karpenter to make an autoscaler?

njtran commented on content/en/docs/concepts/cluster-administration/node-autoscaling.md:
```diff
@@ -0,0 +1,264 @@
+created right after a Node is consolidated, it can get scheduled on an existing Node
+in the cluster instead of the Pods recreated because of consolidation - leaving
+these Pods pending. This scenario is usually automatically recovered by provisioning.
+{{< /note >}}
```
Ah I got it now. 

WDYT about this? 

In consolidating nodes, autoscalers try to simulate how the kube-scheduler will assign pods that are rescheduling. Since autoscalers do not control the actual scheduling process, autoscalers cannot guarantee that "no pods should become pending as a result of consolidation". This can change based on the churn in the cluster, the complexity of scheduling constraints, and other autoscaler-specific constraints.

njtran commented on content/en/docs/concepts/cluster-administration/node-autoscaling.md:
```diff
@@ -0,0 +1,276 @@
+provider. The performance and feature set of a given autoscaler can differ between cloud provider
+integrations.
+
+{{< mermaid >}}
```
I couldn't see this in the rich diff on how this will get created. Can you point me to a link on to what this does?

sftim replied:
https://www.k8s.dev/blog/2021/12/01/improve-your-documentation-with-mermaid.js-diagrams/ explains what's going on here.

sftim replied:
You can also look at the [page preview](https://deploy-preview-45802--kubernetes-io-main-staging.netlify.app/docs/concepts/cluster-administration/node-autoscaling/) @njtran.

njtran replied:
Ah cool, thanks!

sftim commented on content/en/docs/concepts/cluster-administration/node-autoscaling.md:
```diff
@@ -0,0 +1,276 @@
+metadata:
+  name: example
+  annotations:
+    autoscaling.kubernetes.io/do-not-disrupt: true
```
Ideally, register the annotation first
(add an entry into https://kubernetes.io/docs/reference/labels-annotations-taints/)

sftim replied:
Watch out for making a PR that is so big it is harder to review. Splitting into several smaller PRs will help get approval.

sftim commented on content/en/docs/concepts/cluster-administration/node-autoscaling.md:
```diff
@@ -0,0 +1,276 @@
+
+### Configure a workload not to be disrupted by consolidation
+
+{{< feature-state state="stable" for_k8s_version="v1.31" >}}
```
```suggestion
```
This feature was not part of the Kubernetes v1.31 release, because that release hasn't happened, and also because (AIUI) there's no in-tree code that honors this annotation.

sftim replied:
You need to either remove this line, or delay this PR until v1.31 has been released and then rebase.

towca replied:
Understood, I'll remove the section before submitting and follow-up in a separate PR, just wanted some initial feedback from the Karpenter folks on this.

sftim commented on content/en/docs/concepts/cluster-administration/node-autoscaling.md:
```diff
@@ -0,0 +1,276 @@
+
+* Documentation: https://karpenter.sh/
+* Supported cloud providers:
+  https://github.com/kubernetes-sigs/karpenter?tab=readme-ov-file#supported-cloudproviders
```
We should be wary of linking to that section, because the Kubernetes project doesn't formally promise or provide support for either of the things that page links to.

sftim replied:
Karpenter, within Kubernetes, is a library you can use to build a node autoscaler.

njtran replied:
I can't imagine that we'd remove this section, but we might rename it. I think it's fine to link it personally.

jonathan-innis replied:
> because the Kubernetes project doesn't formally promise or provide support for either of the things that page links to

@sftim I'm curious what you mean by this. Are you implying that the K8s project doesn't promise cloudprovider support like this generally?

chrisnegus replied:
Instead of "Karpenter is a multi-cloud project supported on the following cloud providers", couldn't we say something like:

```
Karpenter Implementations
Karpenter is a multi-cloud project with implementations by the following cloud providers:
```

gjtempleton replied:
Agreed with this, can you update the phrasing to @chrisnegus's suggested wording please @towca.

We can come back and update this if we start to see community backed implementations.

sftim replied:
> Are you implying that the K8s project doesn't promise cloudprovider support like this generally?

We support cloud provider integrations if the integration code is part of Kubernetes (for example, we'd triage a bug report and might work on a fix, and security response committee would triage vulnerability reports).

If it's third party - _closed source, or source available but with code outside of the https://github.com/kubernetes/ and https://github.com/kubernetes-sigs/ repositories_ -  we don't promise **any** support. There's an exception for https://github.com/etcd-io/etcd/ which is maintained by a Kubernetes SIG but has its own processes.

sftim replied:
Linking to https://github.com/kubernetes-sigs/karpenter?tab=readme-ov-file#karpenter-implementations is fine.

towca replied:
* Changed "Supported cloud providers" to "Cloud provider integrations" both here and in CA section above.
* The "Implementation comparison" section already covers the differences in how CA and Karpenter integrate with cloud providers.

I think this captures all the suggestions in this thread - WDYT about the current version?

sftim commented on content/en/docs/concepts/cluster-administration/node-autoscaling.md:
```diff
@@ -0,0 +1,276 @@
+* Documentation: https://karpenter.sh/
+* Supported cloud providers:
+  https://github.com/kubernetes-sigs/karpenter?tab=readme-ov-file#supported-cloudproviders
+* Extensive FAQ: https://karpenter.sh/docs/faq/
```
(nit)
```suggestion
* Karpenter FAQ: https://karpenter.sh/docs/faq/
```

sftim commented on content/en/docs/concepts/cluster-administration/node-autoscaling.md:
```diff
@@ -0,0 +1,276 @@
+  https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/README.md
+* Supported cloud providers:
+  https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/README.md#faqdocumentation
+* Extensive FAQ: https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/FAQ.md
```
```suggestion
* Cluster Autoscaler FAQ: https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/FAQ.md
```

sftim commented on content/en/docs/concepts/cluster-administration/node-autoscaling.md:
```diff
@@ -0,0 +1,276 @@
+
+[Horizontal workload autoscaling](/docs/concepts/workloads/autoscaling#scaling-workloads-horizontally)
+automatically scales the number of workload replicas to maintain a desired average resource
+utilization across the replicas. In other words, it automatically creates newPods in response to
```
(nit)
```suggestion
utilization across the replicas. In other words, the control plane automatically creates new Pods in response to
```

sftim commented on content/en/docs/concepts/cluster-administration/node-autoscaling.md:
```diff
@@ -0,0 +1,276 @@
+#### Karpenter
+
+Karpenter auto-provisions VMs based on [NodePool](https://karpenter.sh/docs/concepts/nodepools/)
+configuration provided by the user. Karpenter handles all aspects of node lifecycle, not just
```
(nit)
```
configuration provided by the cluster operator. Karpenter handles all aspects of node lifecycle, not just
```

sftim commented on content/en/docs/concepts/cluster-administration/node-autoscaling.md:
```diff
@@ -0,0 +1,276 @@
+
+Karpenter auto-provisions VMs based on [NodePool](https://karpenter.sh/docs/concepts/nodepools/)
+configuration provided by the user. Karpenter handles all aspects of node lifecycle, not just
+autoscaling. It works directly with VMs and doesn't rely on provider-side VM groups.
```
Karpenter doesn't just work with VMs (it can provision bare metal servers too).

sftim commented on content/en/docs/concepts/cluster-administration/node-autoscaling.md:
```diff
@@ -0,0 +1,276 @@
+From this perspective, the overall Node utilization in a cluster can be used as a proxy for how
+cost-effective the cluster is.
+
+{{< note >}}
+For the purpose of consolidation, resource utilization is computed using Pod resource requests and
+Node allocatable resources. Real resource usage is not taken into account. The details of how
+utilization is computed can vary between autoscalers.
+{{< /note >}}
```
There are a lot of notes.
```suggestion
For the purpose of consolidation, resource utilization is computed using Pod resource requests and
Node allocatable resources. Real resource usage is not taken into account. The details of how
utilization is computed can vary between autoscalers.
```

sftim commented on content/en/docs/concepts/cluster-administration/node-autoscaling.md:
```diff
@@ -0,0 +1,276 @@
+of various limitations (for example reaching configured provisioning limits, provisioning
+configuration not being compatible with a particular set of pods, lack of cloud provider capacity).
+
+{{< note >}}
+The scheduling constraints of pods (for example resource requests, node selectors) are the main
+input when determining Nodes to provision. In particular, the real resource usage of the Pods after
+they start running is not taken into account. In order to autoscale Nodes based on actual resource
+usage, you can [combine workload and Node autoscaling](#multi-scaling).
+{{< /note >}}
```
There are a lot of notes.

Please drop the `note` shortcode, and if you like the rewording, you could try this:
```suggestion
The [scheduling constraints](/docs/concepts/scheduling-eviction/assign-pod-node/) of pods
(for example resource requests, and node selectors) are the main input to a node autoscaler 
when determining whether to provision a new server as a node, and if so what kind of server.
In particular, autoscalers don't directly take into account the real resource usage of the Pods
after they start running. In order to optimize your cluster and its nodes based on actual resource
usage, you can [combine workload and node autoscaling](#multi-scaling).
```

sftim commented on content/en/docs/concepts/cluster-administration/node-autoscaling.md:
```diff
@@ -0,0 +1,276 @@
+
+#### Cluster Autoscaler
+
+Cluster Autoscaler adds or removes Nodes to pre-configured Node groups. Node groups generally map to
```
(nit)
```suggestion
Cluster Autoscaler adds or removes Nodes to pre-configured _node groups_. Node groups generally map to
```

elmiko commented on content/en/docs/concepts/cluster-administration/node-autoscaling.md:
```diff
@@ -0,0 +1,264 @@
+This is especially useful if the number of Pods changes over time, for example
+as a result of [combining workload and Node autoscaling](#multi-scaling).
+
+Autoscalers provision the Nodes by creating and deleting cloud provider Virtual Machines (VMs) backing them.
```
i don't think we need to split it, mainly i just wanted to expand on "virtual machines" to be wider in scope.

were would you split it?

towca replied:
Hmm, have you seen this part after my latest update? I just meant that it's now 2 sentences:

> Autoscalers provision the Nodes by creating and deleting cloud provider resources backing them. Most
commonly, the resources backing the Nodes are Virtual Machines (VMs).

sftim commented:
- this will need a tech review on behalf of @kubernetes/sig-autoscaling-pr-reviews
- @towca, please state what conditions need to be met to unhold the PR
  (in this repo, we prefer to only hold PRs when we know what needs to happen ahead of removing that hold)

towca commented:
> Some more feedback

@sftim Addressed/responded.

> @towca this might be a work in progress. If it is, you can change the PR title to start with “[WIP] ”.

What makes you say that?

> * this will need a tech review on behalf of @kubernetes/sig-autoscaling-pr-reviews

I'd expect so, the relevant reviewers are assigned to the PR. But also interesting, do you have any context of why such tech review wasn't needed in #45197?

> * @towca, please state what conditions need to be met to unhold the PR
>   (in this repo, we prefer to only hold PRs when we know what needs to happen ahead of removing that hold)

Conditions:
* All sig-autoscaling stakeholders (=the PR's assignees) LGTM the PR.

sftim commented:
> But also interesting, do you have any context on why such tech review wasn't needed in https://github.com/kubernetes/website/pull/45197?

I can't speak for those reviewers.

sftim commented on content/en/docs/concepts/cluster-administration/cluster-autoscaling.md:
You should configure a redirect @towca - see `static/_redirects` in this repo. Use a 301 code.

towca replied:
Good point, thanks. Also renamed a bunch of references under `/en/docs/`. Should I also update them in other languages, or are they somehow auto-generated from en?

sftim replied:
Localization teams will pick up these changes; each team defines its own workflow.

sftim commented on content/en/docs/concepts/cluster-administration/node-autoscaling.md:
```diff
@@ -0,0 +1,306 @@
+You can use Node autoscaling together with vertical workload autoscaling in order to adjust the
+resource requests of your Pods while preserving Node autoscaling capabilities in your cluster.
+
+{{< warning >}}
+When using Node autoscaling, it's not recommended to set up vertical workload autoscaling for
+DaemonSet Pods. Autoscalers have to predict how DaemonSet Pods on a new Node will look like in order
+to predict available Node resources. Vertical workload autoscaling might make these predictions
+unreliable, leading to incorrect scaling decisions.
+{{</ warning >}}
```
We'd use `caution` here I think (we == SIG Docs). `warning` is reserved for “pay attention, or your overall information security could be at risk”.

towca replied:
Makes sense, thanks

sftim commented on content/en/docs/concepts/cluster-administration/node-autoscaling.md:
```diff
@@ -0,0 +1,306 @@
+description: >-
+  Automatically provision and consolidate the Nodes in your cluster to adapt to demand and optimize cost.
+content_type: concept
+weight: 130
```
Now that we have https://kubernetes.io/docs/concepts/cluster-administration/node-shutdown/, I suggest
```suggestion
weight: 15
```

sftim commented:
(please rebase this against _main_)

sftim commented on content/en/docs/concepts/cluster-administration/node-autoscaling.md:
```diff
@@ -0,0 +1,306 @@
+determining Nodes to provision. In particular, autoscalers don't directly take into account the real
+resource usage of the Pods after they start running. In order to autoscale Nodes based on actual
+resource usage, you
+can [combine horizontal workload autoscaling with Node autoscaling](#horizontal-workload-autoscaling).
```
Any reviewer for SIG Docs can comment, but yes the suggestion is a good one to make.

sftim commented:
/hold

https://github.com/kubernetes/website/pull/45802#discussion_r1576764436
We shouldn't merge docs for an unregistered annotation.

OK to unhold once that's addressed **and** the other reasons for holding are covered too (@towca wants multiple tech reviews).

jonathan-innis commented on content/en/docs/concepts/cluster-administration/node-autoscaling.md:
```diff
@@ -0,0 +1,306 @@
+#### Karpenter
+
+Karpenter auto-provisions Nodes based on [NodePool](https://karpenter.sh/docs/concepts/nodepools/)
+configuration provided by the cluster operator. Karpenter handles all aspects of node lifecycle, not
```
```suggestion
configurations provided by the cluster operator. Karpenter handles all aspects of node lifecycle, not
```

jonathan-innis commented on content/en/docs/concepts/cluster-administration/node-autoscaling.md:
```diff
@@ -0,0 +1,306 @@
+
+Karpenter auto-provisions Nodes based on [NodePool](https://karpenter.sh/docs/concepts/nodepools/)
+configuration provided by the cluster operator. Karpenter handles all aspects of node lifecycle, not
+just autoscaling. It works directly with individual cloud provider resources (most commonly
```
```suggestion
just autoscaling. This includes automatically refreshing nodes once nodes reach a certain lifetime and auto-upgrading nodes when new worker node images are released. It works directly with individual cloud provider resources (most commonly
```

jonathan-innis commented on content/en/docs/concepts/cluster-administration/node-autoscaling.md:
```diff
@@ -0,0 +1,306 @@
+Karpenter auto-provisions Nodes based on [NodePool](https://karpenter.sh/docs/concepts/nodepools/)
+configuration provided by the cluster operator. Karpenter handles all aspects of node lifecycle, not
+just autoscaling. It works directly with individual cloud provider resources (most commonly
+individual Virtual Machines), and doesn't rely on cloud provider resource groups.
```
I see you call out "most commonly individual VMs" a lot. I wonder if there's reason to call this out because I'm imagining that it's assumed.

jonathan-innis commented on content/en/docs/concepts/cluster-administration/node-autoscaling.md:
```diff
@@ -0,0 +1,306 @@
+
+* Cluster Autoscaler provides features related to just Node autoscaling. Karpenter has a wider
+  scope, and also provides features intended for managing Node lifecycle altogether (for example
+  grouping Nodes into pools, handling Node upgrades via disruption).
```
Not sure how much value there is talking about "grouping nodes into pools". If I were to call out two features to represent the lifecycle behavior, I'd call out rolling nodes when they've reached a certain lifetime (expiry) and auto-upgrade (which you've already called out)

jonathan-innis commented on content/en/docs/concepts/cluster-administration/node-autoscaling.md:
```diff
@@ -0,0 +1,306 @@
+
+## Configure workloads for autoscaling {#workload-config}
+
+Normally, workloads don't have to be additionally configured to work with Node autoscaling, beyond
```
```suggestion
Normally, workloads don't have to be specially configured to work with Node autoscaling, beyond
```

nit

jonathan-innis commented on content/en/docs/concepts/cluster-administration/node-autoscaling.md:
```diff
@@ -0,0 +1,306 @@
+
+{{< feature-state state="stable" for_k8s_version="v1.31" >}}
+
+You can use annotation `autoscaling.kubernetes.io/do-not-disrupt: true` on a Pod to prevent it being
```
Agree with @sftim, it might be worth breaking out the stuff in here that has to do with API alignment so that we don't have to hold this docs change based on the proposed API changes. We obviously still need to discuss this annotation and also get this registered elsewhere in upstream

sftim commented:
@towca my advice here: try to find a cleavage point where you can split this into smaller reviewable PRs. It's a topic area (autoscaling) where docs have historically been OK but not great, which means that most docs reviewers won't be confident reviewing a PR.

The smaller the pool of people who can confidently review a set of changes, and the bigger the change, the longer the lead time on eventual approval.

towca commented on content/en/docs/concepts/cluster-administration/node-autoscaling.md:
```diff
@@ -0,0 +1,306 @@
+Provisioning is formerly known as _scale-up_ in Cluster Autoscaler.
+{{< /note >}}
+
+### Auto-provisioning {#autoprovisioning}
```
There's already some prior discussion on this topic, I wrote down my point of view here: https://github.com/kubernetes/website/pull/45802#discussion_r1576716656. WDYT?

MaciekPytel replied:
I think what this is trying to capture is having to set up ASGs / VMSS / MIGs / whatever else your provider uses for a nodegroup. To me the difference between 'provisioning' and 'auto-provisioning' is that you don't have to do that. It's not easy to define precisely - you still need to set up Kaprenter NodePool or configure GKE NAP, but intuitively there is a major difference in experience between that and having to setup all the "nodegroups" on provider side. 
I wonder if a more clear definition would be saying that auto-provisioning is ability to configure the machines that will be provisioned entirely in a Kubernetes layer (e.g. via CRDs or nodeSelectors)? Admittedly that is not 100% obvious either - what about CAS + Cluster API? But maybe that's fair and CAS + Cluster API is indeed providing this capability that other CAS integrations don't?

sftim replied:
Remember that the perfect is the enemy of the good. With code, it needs to pass all tests before we merge it.

With docs, we believe in rough consensus and factual text. Merging this PR and iterating on it is a more effective way forward.

(I'm a SIG Docs reviewer for English and an emeritus docs tech lead).

towca reviewed (commented):
@jonathan-innis @tallaxes @sftim Addressed/responded to most comments, please let me know if I missed any!

chrisnegus commented on content/en/docs/concepts/cluster-administration/node-autoscaling.md:
```diff
@@ -0,0 +1,289 @@
+to some sort of cloud provider resource group (most commonly a Virtual Machine group). A single
+instance of Cluster Autoscaler can simultaneously manage multiple Node groups. When provisioning,
+Cluster Autoscaler will add Nodes to the group that best fits the requests of pending Pods. When
+consolidating, Cluster Autoscaler always selects specific Nodes to remove as opposed to just
```
```suggestion
consolidating, Cluster Autoscaler always selects specific Nodes to remove, as opposed to just
```

towca replied:
Done

chrisnegus commented on content/en/docs/concepts/cluster-administration/node-autoscaling.md:
```diff
@@ -0,0 +1,289 @@
+Main differences between Cluster Autoscaler and Karpenter:
+
+* Cluster Autoscaler provides features related to just Node autoscaling. Karpenter has a wider
+  scope, and also provides features intended for managing Node lifecycle altogether (for example
```
```suggestion
  scope, and also provides features intended for managing Node lifecycle altogether (for example,
```

towca replied:
Done

chrisnegus commented on content/en/docs/concepts/cluster-administration/node-autoscaling.md:
```diff
@@ -0,0 +1,289 @@
+  them to new versions).
+* Cluster Autoscaler doesn't support auto-provisioning, the Node groups it can provision from have
+  to be pre-configured. Karpenter supports auto-provisioning, so the user only has to configure a
+  set of constraints for the provisioned Nodes instead of fully configuring homogenous groups.
```
```suggestion
  set of constraints for the provisioned Nodes, instead of fully configuring homogenous groups.
```

towca replied:
Done

chrisnegus commented on content/en/docs/concepts/cluster-administration/node-autoscaling.md:
```diff
@@ -0,0 +1,289 @@
+  of the Kubernetes project. For Karpenter, the Kubernetes project publishes Karpenter as a library
+  that cloud providers can integrate with to build a Node autoscaler.
+* Cluster Autoscaler provides integrations with numerous cloud providers, including smaller and less
+  popular providers. There are fewer cloud providers that integrate with Karpenter - including
```
```suggestion
  popular providers. There are fewer cloud providers that integrate with Karpenter, including
```

towca replied:
Done

chrisnegus commented on content/en/docs/concepts/cluster-administration/node-autoscaling.md:
```diff
@@ -0,0 +1,289 @@
+
+### Horizontal workload autoscaling {#horizontal-workload-autoscaling}
+
+Node autoscaling usually works in response to Pods - it provisions new Nodes to accommodate
```
```suggestion
Node autoscaling usually works in response to Pods&mdash;it provisions new Nodes to accommodate
```

towca replied:
Done

chrisnegus commented on content/en/docs/concepts/cluster-administration/node-autoscaling.md:
```diff
@@ -0,0 +1,289 @@
+to accommodate the new Pods.
+
+Once the application load decreases, workload autoscaling should remove unnecessary Pods. Node
+autoscaling should in turn consolidate the Nodes that are no longer needed.
```
```suggestion
autoscaling should, in turn, consolidate the Nodes that are no longer needed.
```

towca replied:
Done

chrisnegus commented on content/en/docs/concepts/cluster-administration/node-autoscaling.md:
```diff
@@ -0,0 +1,289 @@
+
+{{< caution >}}
+When using Node autoscaling, it's not recommended to set up vertical workload autoscaling for
+DaemonSet Pods. Autoscalers have to predict how DaemonSet Pods on a new Node will look like in order
```
```suggestion
DaemonSet Pods. Autoscalers have to predict what DaemonSet Pods on a new Node will look like in order
```

towca replied:
Done

chrisnegus commented on content/en/docs/tasks/run-application/horizontal-pod-autoscale.md:
```diff
@@ -597,7 +597,7 @@ guidelines, which cover this exact use case.
 
 If you configure autoscaling in your cluster, you may also want to consider using
-[cluster autoscaling](/docs/concepts/cluster-administration/cluster-autoscaling/)
+[node autoscaling](/docs/concepts/cluster-administration/node-autoscaling/)
```
```suggestion
[Node Autoscaling](/docs/concepts/cluster-administration/node-autoscaling/)
```

towca replied:
Done

chrisnegus reviewed (commented):
Small wording and punctuation suggestions.

sftim commented on content/en/docs/concepts/cluster-administration/node-autoscaling.md:
```diff
@@ -0,0 +1,289 @@
+
+{{< note >}}
+Autoscalers predict how a recreated Pod will likely be scheduled after a Node is provisioned or
+consolidated, but they don't control the actual scheduling. Because of this, some pods might
```
Lowercase “pods” here is fine according to our style guide (and by me)

towca commented on content/en/docs/concepts/cluster-administration/node-autoscaling.md:
```diff
@@ -0,0 +1,289 @@
+domains).
+
+There are two main inputs to a Node autoscaler when determining Nodes to
+provision - [Pod scheduling constraints](#provisioning-pod-constraints),
```
Done

towca commented on content/en/docs/concepts/cluster-administration/node-autoscaling.md:
```diff
@@ -0,0 +1,289 @@
+Node auto-provisioning is a mode of provisioning in which a user doesn't have to fully configure the
+specifics of the Nodes that can be provisioned. Instead, the autoscaler dynamically chooses the Node
+configuration based on the pending Pods it's reacting to, as well as pre-configured constraints (for
+example the minimum amount of resources, the need for a given label).
```
Done

sftim replied:
I don't see that this is done @towca

towca commented on content/en/docs/concepts/cluster-administration/node-autoscaling.md:
```diff
@@ -0,0 +1,289 @@
+## Node consolidation {#consolidation}
+
+The main component of the cost of running a cluster is usually the cost of the infrastructure (most
+commonly - Virtual Machines) backing its Nodes. In order to minimize this cost and still be able to
```
Done

sftim replied:
I don't see that this is done @towca

towca commented on content/en/docs/concepts/cluster-administration/node-autoscaling.md:
```diff
@@ -0,0 +1,289 @@
+
+For the purpose of consolidation, a Node is considered _empty_ if it only has DaemonSet and static
+Pods running on it. Removing empty Nodes during consolidation is more straightforward than non-empty
+ones, and autoscalers often have optimizations designed specifically for consolidating empty nodes.
```
Done

sftim replied:
I don't see that this is done @towca

towca commented on content/en/docs/concepts/cluster-administration/node-autoscaling.md:
```diff
@@ -0,0 +1,289 @@
+Pods running on it. Removing empty Nodes during consolidation is more straightforward than non-empty
+ones, and autoscalers often have optimizations designed specifically for consolidating empty nodes.
+
+Removing non-empty Nodes during consolidation is disruptive - the Pods running on them are
```
Done

sftim replied:
I don't see that this is done @towca

towca commented on content/en/docs/concepts/cluster-administration/node-autoscaling.md:
```diff
@@ -0,0 +1,289 @@
+appears while consolidation is being performed.
+{{< /note >}}
+
+Autoscaler configuration may also enable triggering consolidation by other conditions (for example
```
Done

sftim replied:
I don't see that this is done @towca

towca commented on content/en/docs/concepts/cluster-administration/node-autoscaling.md:
```diff
@@ -0,0 +1,289 @@
+{{< /note >}}
+
+Autoscaler configuration may also enable triggering consolidation by other conditions (for example
+the time elapsed since a Node was created), in order to optimize different properties (for example
```
Done

sftim replied:
I don't see that this is done @towca

towca commented:
Thanks @chrisnegus, added your suggestions!

sftim reviewed (commented):
This is ready for review. As I understand it @towca you want all the leads from SIG Autoscaling to review this, and also a review from SIG Docs.

Have I got that right?

towca commented:
@sftim Yes that's right.

@gjtempleton @jonathan-innis @MaciekPytel The review has stalled for some time, could you please prioritize?

gjtempleton commented on content/en/docs/concepts/cluster-administration/node-autoscaling.md:
```diff
@@ -0,0 +1,289 @@
+The most common kind of scheduling constraints are the resource requests specified by Pod
+containers. Autoscalers will make sure that the provisioned Nodes have enough resources to satisfy
+the requests. However, they don't directly take into account the real resource usage of the Pods
+after they start running. In order to autoscale Nodes based on actual resource usage, you can
+combine [horizontal workload autoscaling](#horizontal-workload-autoscaling) with Node autoscaling.
```
This sentence slightly worries me that people will read `based on actual resource usage` as the classic ASG example of scaling on CPU usage of the VMs. Maybe make clear that it's purely workload resource usage here?

sftim replied:
I don't see that this is done @towca

gjtempleton commented on content/en/docs/concepts/cluster-administration/node-autoscaling.md:
```diff
@@ -0,0 +1,306 @@
+
+## Node consolidation {#consolidation}
+
+The main component of the cost of running a cluster is usually the cost of the infrastructure (most
```
Agreed with Jonathan here. Maybe something along the lines of 

```
The main consideration when running a cluster is ensuring that all schedulable pods are running, 
whilst keeping the cost of the cluster as low as possible. To achieve this the Pods' resource 
requests should utilize as much of the Nodes' resources as possible. From this perspective, 
the overall Node utilization in a cluster can be used as a proxy for how cost-effective the 
cluster is.
```
?

sftim replied:
I don't see that this is changed @towca

gjtempleton commented on content/en/docs/concepts/cluster-administration/node-autoscaling.md:
```diff
@@ -0,0 +1,289 @@
+{{< note >}}
+Autoscalers predict how a recreated Pod will likely be scheduled after a Node is provisioned or
+consolidated, but they don't control the actual scheduling. Because of this, some Pods might
+technically become pending as a result of consolidation - if for example a completely new Pod
```
```suggestion
become pending as a result of consolidation - if for example a completely new Pod
```

Would drop technically - non-zero chance of this actually happening.

towca replied:
One might argue that in this scenario the pods became pending as a result of new pods appearing, not the initial consolidation - hence the technically. It seems that it's confusing without this further clarification though, dropping it sgtm

gjtempleton commented on content/en/docs/concepts/cluster-administration/node-autoscaling.md:
```diff
@@ -0,0 +1,289 @@
+the maximum lifespan of Nodes in a cluster).
+
+The details of how consolidation is performed depend on the configuration of a given autoscaler.
+Consult [Configure workloads for autoscaling](#workload-config) for common workload-level
```
This appears to be a dead link

towca replied:
Thanks for the catch! It refers to a section that was in the initial PR but was removed during the review. Removed

gjtempleton commented on content/en/docs/concepts/cluster-administration/node-autoscaling.md:
```diff
@@ -0,0 +1,289 @@
+  of the Kubernetes project. For Karpenter, the Kubernetes project publishes Karpenter as a library
+  that cloud providers can integrate with to build a Node autoscaler.
+* Cluster Autoscaler provides integrations with numerous cloud providers, including smaller and less
+  popular providers. There are fewer cloud providers that integrate with Karpenter, including
+  [AWS](https://github.com/aws/karpenter-provider-aws), and
+  [Azure](https://github.com/Azure/karpenter-provider-azure).
```
```suggestion
  popular providers. There are currently fewer cloud providers with Karpenter integrations, including
  [AWS](https://github.com/aws/karpenter-provider-aws), and
  [Azure](https://github.com/Azure/karpenter-provider-azure). See the project documentation for a [comprehensive list](https://github.com/kubernetes-sigs/karpenter?tab=readme-ov-file#supported-cloudproviders).
```

I see from a [previous comment](https://github.com/kubernetes/website/pull/45802#discussion_r1588897225) that you had concerns on referencing these implementations @sftim, how would you feel about the above phrasing?

sftim replied:
We typically avoid “currently” in documentation. I'm also wary of “comprehensive”; how confident we're sure that the adjective will stay true long term?

sftim replied:
Also, we should link to https://github.com/kubernetes-sigs/karpenter?tab=readme-ov-file#karpenter-implementations (the README has changed).

towca replied:
Is the current wording (no "currently", no "comprehensive") ok with you @sftim?

gjtempleton reviewed (commented):
A few small nits from me, but this mostly looks like a massive improvement, great stuff.

sftim commented on content/en/docs/concepts/cluster-administration/node-autoscaling.md:
```diff
@@ -0,0 +1,289 @@
+
+* Documentation: https://karpenter.sh/
+* Supported cloud providers:
+  https://github.com/kubernetes-sigs/karpenter?tab=readme-ov-file#supported-cloudproviders
```
```suggestion
  https://github.com/kubernetes-sigs/karpenter?tab=readme-ov-file#karpenter-implementations
```

towca replied:
Done

MaciekPytel commented on content/en/docs/concepts/cluster-administration/node-autoscaling.md:
```diff
@@ -0,0 +1,287 @@
+can combine [horizontal workload autoscaling](#horizontal-workload-autoscaling) with Node
+autoscaling.
+
+Other common Pod scheduling constraints include Node affinity, inter-Pod affinity, or a requirement
```
nit: I wonder if we want to add links to node affinity and pod affinity here?

MaciekPytel commented:
I left some minor comments, but overall I think this looks good. I think we need to keep the hold until @gjtempleton and @jonathan-innis lgtm, but from my side:

/lgtm
/approve

gjtempleton commented:
Looks good for me now as well.

towca commented:
@jonathan-innis We have an LGTM from the CA side now - just missing the Karpenter LGTM and a final pass from the docs approvers. I addressed all of your comments, could we close on this soon?

sftim commented on content/en/docs/concepts/cluster-administration/node-autoscaling.md:
```diff
@@ -0,0 +1,292 @@
+* Documentation: https://karpenter.sh/
+* Supported cloud providers:
+  https://github.com/kubernetes-sigs/karpenter?tab=readme-ov-file#supported-cloudproviders
+* Karpenter FAQ: https://karpenter.sh/docs/faq/
+* Contact:
+  https://github.com/kubernetes-sigs/karpenter#community-discussion-contribution-and-support
```
Please use real hyperlinks with anchor text.

sftim commented on content/en/docs/concepts/cluster-administration/node-autoscaling.md:
```diff
@@ -0,0 +1,292 @@
+
+Additional context:
+
+* Documentation overview:
+  https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/README.md
+* Supported cloud providers:
+  https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/README.md#faqdocumentation
+* Cluster Autoscaler FAQ:
+  https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/FAQ.md
+* Contact: https://github.com/kubernetes/community/tree/master/sig-autoscaling#contact
```
Please use real hyperlinks with anchor text.

sftim commented on content/en/docs/concepts/cluster-administration/node-autoscaling.md:
```diff
@@ -0,0 +1,292 @@
+
+Additional context:
+
+* Documentation: https://karpenter.sh/
```
Please use a real hyperlink, with anchor text.

sftim commented on content/en/docs/concepts/cluster-administration/node-autoscaling.md:
```diff
@@ -0,0 +1,292 @@
+configuration options.
+
+{{< note >}}
+Consolidation is formerly known as _scale-down_ in Cluster Autoscaler.
```
(nit)
```suggestion
Consolidation was formerly known as _scale-down_ in Cluster Autoscaler.
```

sftim commented on content/en/docs/concepts/cluster-administration/node-autoscaling.md:
```diff
@@ -0,0 +1,289 @@
+because of various limitations, including reaching configured provisioning limits, provisioning
+configuration not being compatible with a particular set of pods, or the lack of cloud provider
+capacity. While provisioning, Node autoscalers often try to achieve additional goals (for example
+minimizing the cost of the provisioned Nodes, or balancing the number of Nodes between failure
```
I don't see that this is done @towca

sftim commented on content/en/docs/concepts/cluster-administration/node-autoscaling.md:
```diff
@@ -0,0 +1,289 @@
+## Node provisioning {#provisioning}
+
+If there are Pods in a cluster that can't be scheduled on existing Nodes, new Nodes can be
+automatically added to the cluster - _provisioned_ - to accommodate the Pods. This is especially
```
I don't see that this is done @towca

sftim commented:
There's quite a lot of pending feedback, too. See https://github.com/kubernetes/website/pull/45802/files.

towca commented:
@sftim Ugh, I messed up the push and scrapped the latest round of changes. Sorry for that, they should be restored now

sftim commented:
This is OK to unhold once tech LGTM(s) are in place from SIG Autoscaling.

tengqm commented:
Why are we changing "cluster autoscaling" to "node autoscaling"?
My (previous and current) understanding is that cluster scaling deals with the scale of a cluster, while node scaling means adding/removing resources to/from a node. This could be wrong although I believe many users may get confused as I do.

sftim commented:
> Why are we changing "cluster autoscaling" to "node autoscaling"?

SIG Autoscaling had a brief to do this; see https://github.com/kubernetes/autoscaler/issues/6646 for more context.

I don't object to that change. Horizontally autoscaling your nodes is equivalent to horizontally autoscaling your VMs, Pods, etc. Or you can use a multi dimensional autoscaler like Karpenter.

towca commented:
@jonathan-innis @njtran Ping, could we close on this review? It's been open forever, and I think we've reached a consensus on the controversial parts now.

sftim commented on content/en/docs/concepts/cluster-administration/node-autoscaling.md:
:thought_balloon: We could mention that your cloud provider may provide managed node autoscaling, and to consult your cloud provider's docs for more details.

sftim commented:
@towca you don't have to get a review from SIG Autoscaling on this. The _required_ review and approval is from SIG Docs.

sftim commented:
If you're happy to have SIG Docs take the reviewing forward, reply to confirm that. We can get a technical review using our usual approach.

jonathan-innis commented on content/en/docs/concepts/cluster-administration/node-autoscaling.md:
```diff
@@ -0,0 +1,282 @@
+## Node consolidation {#consolidation}
+
+The main consideration when running a cluster is ensuring that all schedulable pods are running,
+whilst keeping the cost of the cluster as low as possible. To achieve this the Pods' resource
```
```suggestion
whilst keeping the cost of the cluster as low as possible. To achieve this, the Pods' resource
```

nit

towca replied:
Done

jonathan-innis commented on content/en/docs/concepts/cluster-administration/node-autoscaling.md:
```diff
@@ -0,0 +1,282 @@
+  of the Kubernetes project. For Karpenter, the Kubernetes project publishes Karpenter as a library
+  that cloud providers can integrate with to build a Node autoscaler.
+* Cluster Autoscaler provides integrations with numerous cloud providers, including smaller and less
+  popular providers. There are fewer cloud providers that integrate with Karpenter, including
```
There's more of these now including https://github.com/cloudpilot-ai/karpenter-provider-alibabacloud and https://github.com/kubernetes-sigs/karpenter-provider-cluster-api. We may end up continually updating this list but I'm also fine taking this as a follow-on

towca replied:
I'll take you up on the offer then, I'd prefer not to delay this PR further.

jonathan-innis reviewed (approved):
/lgtm

Thanks for putting this together @towca! Sorry, this took so long, but I think this is going to be a huge benefit to everyone once this gets in!

jonathan-innis commented:
/lgtm

sftim commented:
> The PR is ready for SIG Docs review/approval!

I have just the GIF

![Ship it squirrel](https://github.com/user-attachments/assets/0ec73fd5-90ac-457b-a742-849c798554a1)

/approve
